### PR TITLE
Support `col[span]` and `pt`s in paste from excel

### LINKED ISSUE #14516: Tables pasted from Excel get bogus extra spacing/padding on Chrome
## 📝 Provide detailed reproduction steps (if any)

1.  Download [Excel (3).xlsx](https://github.com/ckeditor/ckeditor5/files/11935759/Excel.3.xlsx) file & open it in Excel.
2.  Select last three columns, like so:

1.  Paste it into https://ckeditor.com/docs/ckeditor5/latest/examples/builds-custom/full-featured-editor.html

### ✔️ Expected result

Cell widths are preserved (cell with "color on text and row" is wider than the other two).

### ❌ Actual result

Extra spacing is visible at the right hand side.

## 📃 Other details

Regression in 38.1.0.

Not reproducible on Firefox.

*   Browser: …
*   OS: …
*   First affected CKEditor version: …
*   Installed CKEditor plugins: …

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### LINKED ISSUE #14521: Columns width for tables pasted from MS Excel are not conveyed properly
## 📝 Provide detailed reproduction steps (if any)

1.  Download [Excel (3).xlsx](https://github.com/ckeditor/ckeditor5/files/11935759/Excel.3.xlsx) file & open it in Excel.
2.  Select last three columns, like so:

1.  Paste it into [https://ckeditor.com/docs/ckeditor5/latest/examples/builds-custom/full-featured-editor.html](https://ckeditor.com/docs/ckeditor5/latest/examples/builds-custom/full-featured-editor.html)

### ✔️ Expected result

Cell widths are preserved (cell with "color on text and row" is wider than the other two).

Example, see [34.2.0 full featured editor example](https://ckeditor.com/docs/ckeditor5/34.2.0/examples/builds-custom/full-featured-editor.html):

_Note it doesn't retain styles, but that's not the scope of this issue. We care only about widths here._

### ❌ Actual result

Depending on version results were bad over time.

With column resize enabled it started to set wrong column widths, see [table column resize demo in **35.0.1**](https://ckeditor.com/docs/ckeditor5/35.0.1/features/table.html#table-column-resize):

[**38.1.0**](https://ckeditor.com/docs/ckeditor5/38.1.0/examples/builds-custom/full-featured-editor.html) due to #14516  it looks even worse:

## 📃 Other details

*   Browser: Any
*   OS: Any
*   First affected CKEditor version: 35.0.1
*   Installed CKEditor plugins: TableColumnResize

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by Dumluregn**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Fix (table,paste-from-office): Tables pasted from MS Excel will have proper column widths. Closes #14521. Closes #14516.

---

### Additional information

See the detailed explanation what is going on [here](https://github.com/ckeditor/ckeditor5/issues/14521#issuecomment-1662102889).

There's also another PR that should be merged after this one: https://github.com/cksource/ckeditor5-commercial/pull/5399.

### CODE REVIEW

**Comment by dufipl** [comment]
This case is still not working: https://github.com/ckeditor/ckeditor5/issues/14521#issuecomment-1663486990

**Comment by illia-stv** [comment]
CI failing

File: packages/ckeditor5-table/src/tableclipboard.ts [hunk 1]
```diff
@@ -150,6 +150,8 @@ export default class TableClipboard extends Plugin {
 			return;
 		}
 
+		console.log( pastedTable );
```
Comment (illia-stv) [inline packages/ckeditor5-table/src/tableclipboard.ts:153]: `console.log()` should be removed

**Review by illia-stv — APPROVED** [review]
For me looks good✅

File: packages/ckeditor5-table/src/tablecolumnresize/utils.ts [hunk 2]
```diff
@@ -376,23 +373,55 @@ export function getColumnGroupElement( element: Element ): Element {
...
  * @internal
  * @param element A 'table' or 'tableColumnGroup' element.
  * @returns An array of 'tableColumn' elements.
  */
 export function getTableColumnElements( element: Element ): Array<Element> {
-	return Array.from( getColumnGroupElement( element ).getChildren() as IterableIterator<Element> );
+	const columnGroupElement = getColumnGroupElement( element );
+
+	if ( !columnGroupElement ) {
+		return [];
```
[the remark this answers is not in the document]
Comment (mlewand) [reply packages/ckeditor5-table/src/tablecolumnresize/utils.ts:386]: This change should be covered with unit tests.

~While at it I noticed that~ [~test suite for this function has a wrong name~](https://github.com/ckeditor/ckeditor5/blob/e3216289ddbd8d2f7ef7b5433a51f9601639ef05/packages/ckeditor5-table/tests/tablecolumnresize/utils.js#L667) ~as it should be `getTableColumnsWidths()`. Please correct it while we're at it 🙏~ Scratch that, I can see that you also noticed it and that's already changed :) good one.

File: packages/ckeditor5-table/tests/tablecolumnresize/tablecolumnresizeediting.js [hunk 3]
```diff
@@ -482,6 +566,109 @@ describe( 'TableColumnResizeEditing', () => {
 						'</table>'
 					);
 				} );
+
+				it( 'should consume the `col[span]` attribute', () => {
```
Comment (mlewand) [inline packages/ckeditor5-table/tests/tablecolumnresize/tablecolumnresizeediting.js:570]:
```suggestion
				it( 'should convert the `col[span]` attribute', () => {
```

This unit test doesn't check whether the conversion was consumed or not.

**Review by mlewand — COMMENTED** [review]
Good solution, just few minor remarks and we're good to go.

**Review by mlewand — APPROVED** [review]
LGTM, let's merge it once CI passes.

File: packages/ckeditor5-table/src/tablecolumnresize/converters.ts [hunk 4]
```diff
@@ -38,9 +38,10 @@ export function upcastColgroupElement( tableUtilsPlugin: TableUtils ): ( dispatc
...
 
 		const columnElements = getTableColumnElements( tableColumnGroup );
-		let columnWidths = getTableColumnsWidths( tableColumnGroup );
 		const columnsCount = tableUtilsPlugin.getColumns( modelTable );
+		let columnWidths = translateColSpanAttribute( tableColumnGroup, conversionApi.writer );
 
+		// Fill the array with 'auto' values if the number of columns is higher than number of declared values.
 		columnWidths = Array.from( { length: columnsCount }, ( _, index ) => columnWidths[ index ] || 'auto' );
 
 		if ( columnWidths.length != columnElements.length || columnWidths.includes( 'auto' ) ) {
```
Comment (niegowski) [inline packages/ckeditor5-table/src/tablecolumnresize/converters.ts:47]: This is not triggering column width normalization for such case:
```html
<table>
    <colgroup>
        <col style="width: 30pt;">
        <col style="width: 70pt;">
    </colgroup>
    <tbody>
        <tr>
            <td>a</td>
            <td>b</td>
        </tr>
    </tbody>
</table>
```
And the final table ends up like this:
```html
<figure class="table">
    <table class="ck-table-resized">
        <colgroup>
            <col style="width:30%;">
            <col style="width:70%;">
        </colgroup>
        <tbody>
        <tr>
            <td>a</td>
            <td>b</td>
        </tr>
        </tbody>
    </table>
</figure>
```
so `pt` values are treated as they would be `%`, the table gets full page width and the original sizes are not preserved.

The column width is normalized in post-fixer (but should be normalized before post-fixers are triggered).

**Pull request merged**